receiver/awscontainerinsightreceiver/internal/k8swindows/kubelet.go

```diff
@@ -13,82 +13,182 @@ import (
+
 	"go.uber.org/zap"
 	stats "k8s.io/kubelet/pkg/apis/stats/v1alpha1"
 )
 
-type kubeletSummaryProvider struct {
+// KubeletProvider Represents interface to kubelet.
+type KubeletProvider interface {
+	GetClient() (*kubeletutil.KubeletClient, error)
+	GetSummary() (*stats.Summary, error)
+}
+
+type kubeletProvider struct {
 	logger   *zap.Logger
 	hostIP   string
 	hostPort string
 	client   *kubeletutil.KubeletClient
-	hostInfo host.Info
 }
 
-func new(logger *zap.Logger, info host.Info) (*kubeletSummaryProvider, error) {
-	hostIP := os.Getenv("HOST_IP")
-	kclient, err := kubeletutil.NewKubeletClient(hostIP, ci.KubeSecurePort, logger)
+// GetClient Returns singleton kubelet client.
+func (k *kubeletProvider) GetClient() (*kubeletutil.KubeletClient, error) {
+	if k.client != nil {
+		return k.client, nil
+	}
+	kclient, err := kubeletutil.NewKubeletClient(k.hostIP, k.hostPort, k.logger)
+	if err != nil {
+		k.logger.Error("failed to initialize kubeletProvider client, ", zap.Error(err))
+		return nil, err
+	}
+	k.client = kclient
+	return kclient, nil
+}
+
+// GetSummary Get Summary from kubelet API.
+func (k *kubeletProvider) GetSummary() (*stats.Summary, error) {
+	kclient, err := k.GetClient()
 	if err != nil {
-		return nil, fmt.Errorf("failed to initialize kubelet client: %w", err)
+		k.logger.Error("failed to get kubeletProvider client, ", zap.Error(err))
+		return nil, err
+	}
+
+	summary, err := kclient.Summary(k.logger)
+	if err != nil {
+		k.logger.Error("kubeletProvider summary API failed, ", zap.Error(err))
+		return nil, err
+	}
+	return summary, nil
+}
+
+func createDefaultKubeletProvider(logger *zap.Logger) KubeletProvider {
+	kSP := &kubeletProvider{logger: logger, hostIP: os.Getenv("HOST_IP"), hostPort: ci.KubeSecurePort}
+	return kSP
+}
+
+// kubelet represents receiver to get metrics from kubelet.
+type kubelet struct {
+	logger          *zap.Logger
+	kubeletProvider KubeletProvider
+	hostInfo        cExtractor.CPUMemInfoProvider
+}
+
+// options decorates kubelet struct.
+type options func(provider *kubelet)
+
+func new(logger *zap.Logger, info cExtractor.CPUMemInfoProvider, opts ...options) (*kubelet, error) {
+	kSP := &kubelet{
+		logger:          logger,
+		hostInfo:        info,
+		kubeletProvider: createDefaultKubeletProvider(logger),
 	}
 
-	return &kubeletSummaryProvider{
-		logger:   logger,
-		client:   kclient,
-		hostInfo: info,
-	}, nil
+	for _, opt := range opts {
+		opt(kSP)
+	}
+	return kSP, nil
 }
 
-func (k *kubeletSummaryProvider) getMetrics() ([]*cExtractor.CAdvisorMetric, error) {
-	summary, err := k.client.Summary(k.logger)
+func (k *kubelet) getMetrics() ([]*cExtractor.CAdvisorMetric, error) {
+	var metrics []*cExtractor.CAdvisorMetric
+
+	summary, err := k.kubeletProvider.GetSummary()
 	if err != nil {
-		k.logger.Error("kubelet summary API failed, ", zap.Error(err))
+		k.logger.Error("kubeletProvider summary API failed, ", zap.Error(err))
```

Comment:
inconsistent naming.  kubelet or kubeletProvider?

Reply:
it should be kubeletProvider.
I have 2 different structs related to kubelet
1. `kubeletProvider`
2. `SummaryProvider`